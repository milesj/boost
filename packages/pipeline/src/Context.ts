import { isObject, isPlainObject } from '@boost/common';

interface Cloneable {
  clone?: () => unknown;
}

export default class Context {
  /**
   * Create a new instance of the current context and shallow clone all properties.
   */
  clone(...args: any[]): this {
    // @ts-expect-error
    const context = new this.constructor(...args);

    // Copy enumerable properties
    Object.keys(this).forEach((key) => {
      const prop = key as keyof this;
      let value: unknown = this[prop];

      if (Array.isArray(value)) {
        value = [...value];
      } else if (value instanceof Map) {
        value = new Map(value);
      } else if (value instanceof Set) {
        value = new Set(value);
      } else if (value instanceof Date) {
        value = new Date(value.getTime());
      } else if (isObject<Cloneable>(value)) {
        if (typeof value.clone === 'function') {
          value = value.clone();
          // Dont dereference instances, only plain objects
        } else if (isPlainObject(value)) {
          value = { ...value };
        }
      }

      context[prop] = value;
    });

    return context;
  }
}
