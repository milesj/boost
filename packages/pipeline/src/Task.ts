import { Blueprint, Predicates } from '@boost/common';
import WorkUnit from './WorkUnit';

export default class Task<Input = unknown, Output = Input> extends WorkUnit<{}, Input, Output> {
	// A task is simply a work unit that executes a standard function.
	// It doesn't need configurable options, so implement an empty blueprint.
	blueprint(preds: Predicates): Blueprint<object> {
		return {};
	}
}
