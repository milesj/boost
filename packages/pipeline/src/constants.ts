import { createInternalDebugger } from '@boost/internal';
import { Status } from './types';

export const debug = createInternalDebugger('pipeline');

export const STATUS_PENDING: Status = 'pending';
export const STATUS_RUNNING: Status = 'running';
export const STATUS_SKIPPED: Status = 'skipped';
export const STATUS_PASSED: Status = 'passed';
export const STATUS_FAILED: Status = 'failed';
