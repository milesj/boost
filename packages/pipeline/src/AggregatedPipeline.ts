import ParallelPipeline from './ParallelPipeline';
import Context from './Context';
import { AggregatedResult } from './types';

export default class AggregatedPipeline<
  Ctx extends Context,
  Input,
  Output = Input
> extends ParallelPipeline<{}, Ctx, Input, Output> {
  blueprint() {
    return {};
  }

  /**
   * Execute all work units in parallel with a value being passed to each work unit.
   * Work units will synchronize regardless of race conditions and errors.
   */
  async run(): Promise<AggregatedResult<Output>> {
    const { context, value } = this;
    const work = this.getWorkUnits();

    this.debug('Aggregating %d work units', work.length);
    this.onRun.emit([value]);

    return Promise.all(
      work.map(unit => {
        this.onRunWorkUnit.emit([unit, value]);

        return unit.run(context, value).catch(error => error);
      }),
    ).then(responses => this.aggregateResult(responses));
  }
}
