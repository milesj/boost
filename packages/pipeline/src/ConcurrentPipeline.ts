import ParallelPipeline from './ParallelPipeline';
import Context from './Context';
import debug from './debug';

export default class ConcurrentPipeline<
  Ctx extends Context,
  Input = unknown,
  Output = Input
> extends ParallelPipeline<{}, Ctx, Input, Output> {
  blueprint() {
    return {};
  }

  /**
   * Execute all work units in parallel with a value being passed to each work unit.
   * If an error occurs, the pipeline will abort early, otherwise return a list of all results.
   */
  async run(): Promise<Output[]> {
    const { context, value } = this;
    const work = this.getWorkUnits();

    debug('Running %d in parallel', work.length);

    this.onRun.emit([value]);

    const result = await Promise.all(
      work.map((unit) => {
        this.onRunWorkUnit.emit([unit, value]);

        return unit.run(context, value);
      }),
    );

    this.onFinish.emit([]);

    return result;
  }
}
