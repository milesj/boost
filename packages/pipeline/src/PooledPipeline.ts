/* eslint-disable promise/prefer-await-to-callbacks */
/* eslint-disable promise/prefer-await-to-then */

import os from 'os';
import { Blueprint, Schemas } from '@boost/common';
import { Context } from './Context';
import { debug } from './debug';
import { ParallelPipeline } from './ParallelPipeline';
import { PipelineError } from './PipelineError';
import { AggregatedResult } from './types';
import { WorkUnit } from './WorkUnit';

export interface PooledOptions {
	/** How many work units to process in parallel. */
	concurrency?: number;
	/** Process with first-in-last-out order, instead of first-in-first-out. */
	filo?: boolean;
	/** Timeout in milliseconds that each work unit may run, or `0` to avoid a
  timeout. Defaults to `0`. */
	timeout?: number;
}

export class PooledPipeline<
	Ctx extends Context,
	Input = unknown,
	Output = Input,
> extends ParallelPipeline<PooledOptions, Ctx, Input, Output> {
	protected resolver?: (response: AggregatedResult<Output>) => void;

	protected results: (Error | Output)[] = [];

	protected running: WorkUnit<{}, Input, Output>[] = [];

	override blueprint(schemas: Schemas): Blueprint<PooledOptions> {
		const { bool, number } = schemas;

		return {
			concurrency: number(os.cpus().length).gte(1),
			filo: bool(),
			timeout: number(0).gte(0),
		};
	}

	/**
	 * Execute all work units in parallel, in a pool with a max concurrency,
	 * with a value being passed to each work unit.
	 * Work units will synchronize regardless of race conditions and errors.
	 */
	async run(): Promise<AggregatedResult<Output>> {
		debug('Running %d as a pool', this.work.length);

		this.onBeforeRun.emit([this.value]);

		const result = await new Promise<AggregatedResult<Output>>((resolve) => {
			if (this.work.length === 0) {
				resolve(this.aggregateResult([]));

				return;
			}

			this.resolver = resolve;

			void Promise.all(
				this.work
					.slice(0, this.options.concurrency)
					.map(() => this.runWorkUnit(this.context, this.value)),
			);
		});

		this.onAfterRun.emit([]);

		return result;
	}

	/**
	 * Run a single work unit from the queue, and start the next work unit when it passes or fails.
	 */
	protected async runWorkUnit(context: Ctx, value: Input): Promise<void> {
		const { concurrency, filo, timeout } = this.options;
		const unit = filo ? this.work.pop()! : this.work.shift()!;

		this.running.push(unit);

		const handleResult = (result: Error | Output) => {
			this.running = this.running.filter((running) => running !== unit);
			this.results.push(result);

			if (this.work.length > 0 && this.running.length < concurrency) {
				return this.runWorkUnit(context, value);
			}
			if (this.work.length === 0 && this.running.length === 0 && this.resolver) {
				this.resolver(this.aggregateResult(this.results));
			}

			return Promise.resolve();
		};

		return new Promise((resolve) => {
			let timer: NodeJS.Timeout;

			if (timeout > 0) {
				timer = setTimeout(() => {
					resolve(handleResult(new PipelineError('WORK_TIME_OUT')));
				}, timeout);
			}

			this.onRunWorkUnit.emit([unit, value]);

			unit
				.run(context, value)
				.then((result) => {
					if (timer) {
						clearTimeout(timer);
					}

					resolve(handleResult(result));

					return result;
				})
				.catch((error) => {
					resolve(handleResult(error as Error));
				});
		});
	}
}
