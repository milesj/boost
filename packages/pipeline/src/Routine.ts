import execa, { Options as ExecaOptions, ExecaChildProcess } from 'execa';
import kebabCase from 'lodash/kebabCase';
import split from 'split';
import { createDebugger, Debugger } from '@boost/debug';
import { Event } from '@boost/event';
import Context from './Context';
import WorkUnit from './WorkUnit';
import ConcurrentPipeline from './ConcurrentPipeline';
import PooledPipeline, { PooledOptions } from './PooledPipeline';
import AggregatedPipeline from './AggregatedPipeline';
import WaterfallPipeline from './WaterfallPipeline';

export interface ExecuteCommandOptions {
  workUnit?: WorkUnit<any, any, any>;
}

export default abstract class Routine<
  Options extends object,
  Input,
  Output = Input
> extends WorkUnit<Options, Input, Output> {
  readonly debug: Debugger;

  readonly key: string;

  // Emits before the command is ran
  readonly onCommand = new Event<[string]>('command');

  // Emits on each line chunk of the running command
  readonly onCommandData = new Event<[string, string]>('command-data');

  constructor(key: string, title: string, options?: Options) {
    super(title, (context, value) => this.execute(context, value), options);

    if (!key || typeof key !== 'string') {
      throw new Error('Routine key must be a valid unique string.');
    }

    this.key = kebabCase(key);
    this.debug = createDebugger(['routine', this.key]);
  }

  /**
   * Execute a command with the given arguments and pass the results through a promise.
   */
  async executeCommand(
    command: string,
    args: string[],
    options: ExecaOptions & ExecuteCommandOptions = {},
  ): Promise<ExecaChildProcess> {
    const { workUnit, ...opts } = options;
    const stream = execa(command, args, opts);

    this.onCommand.emit([command]);

    // Push chunks to the reporter
    const unit = workUnit || this;
    const handler = (line: string) => {
      if (unit.isRunning()) {
        unit.output += line;

        // Only capture the status when not empty
        if (line) {
          unit.statusText = line;
        }

        this.onCommandData.emit([command, line]);
      }
    };

    stream.stderr!.pipe(split()).on('data', handler);
    stream.stdout!.pipe(split()).on('data', handler);

    return stream as any;
  }

  /**
   * Create and return a `AggregatedPipeline`. This pipeline will execute all work units
   * in parallel without interruption. Returns a list of errors and results once all resolve.
   */
  createAggregatedPipeline<C extends Context, I, O = I>(context: C, value: I) {
    return new AggregatedPipeline<C, I, O>(context, value);
  }

  /**
   * Create and return a `ConcurrentPipeline`. This pipeline will execute all work units
   * in parallel. Returns a list of values once all resolve.
   */
  createConcurrentPipeline<C extends Context, I, O = I>(context: C, value: I) {
    return new ConcurrentPipeline<C, I, O>(context, value);
  }

  /**
   * Create and return a `PooledPipeline`. This pipeline will execute a distinct set of work units
   * in parallel without interruption, based on a max concurrency, until all work units have ran.
   * Returns a list of errors and results once all resolve.
   */
  createPooledPipeline<C extends Context, I, O = I>(context: C, value: I, options?: PooledOptions) {
    return new PooledPipeline<C, I, O>(context, value, options);
  }

  /**
   * Create and return a `WaterfallPipeline`. This pipeline will execute each work unit one by one,
   * with the return value of the previous being passed to the next. Returns the final value once
   * all resolve.
   */
  createWaterfallPipeline<C extends Context, I>(context: C, value: I) {
    return new WaterfallPipeline<C, I>(context, value);
  }

  /**
   * Execute the current routine and return a new value.
   */
  abstract async execute<Ctx extends Context>(context: Ctx, value: Input): Promise<Output>;
}
