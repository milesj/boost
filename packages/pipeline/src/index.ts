/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import AsyncPipeline from './AsyncPipeline';
import ConcurrentPipeline from './ConcurrentPipeline';
import Context from './Context';
import PooledPipeline, { PooledOptions } from './PooledPipeline';
import Routine from './Routine';
import SyncPipeline from './SyncPipeline';
import SynchronizedPipeline from './SynchronizedPipeline';
import Task from './Task';
import WaterfallPipeline from './WaterfallPipeline';
import WorkUnit from './WorkUnit';

export * from './constants';
export * from './types';

export {
  AsyncPipeline,
  ConcurrentPipeline,
  Context,
  PooledPipeline,
  PooledOptions,
  Routine,
  SyncPipeline,
  SynchronizedPipeline,
  Task,
  WaterfallPipeline,
  WorkUnit,
};
