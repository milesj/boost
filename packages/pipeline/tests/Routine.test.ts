import execa from 'execa';
import { Predicates } from '@boost/common';
import Routine from '../src/Routine';
import Context from '../src/Context';
import AggregatedPipeline from '../src/AggregatedPipeline';
import ConcurrentPipeline from '../src/ConcurrentPipeline';
import PooledPipeline from '../src/PooledPipeline';
import WaterfallPipeline from '../src/WaterfallPipeline';
import Task from '../src/Task';
import { STATUS_RUNNING } from '../src/constants';

jest.mock('execa');

describe('Routine', () => {
  class TestRoutine extends Routine<{ test: number }, any, any> {
    blueprint({ number }: Predicates) {
      return {
        test: number(),
      };
    }

    async execute() {
      return null;
    }
  }

  let routine: TestRoutine;

  beforeEach(() => {
    routine = new TestRoutine('key', 'title');
  });

  it('errors if no key is provided', () => {
    expect(() => new TestRoutine('', 'title')).toThrowErrorMatchingSnapshot();
  });

  it('errors if key is not a string', () => {
    expect(
      () =>
        new TestRoutine(
          // @ts-ignore Allow invalid type
          123,
          'title',
        ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('inherits options', () => {
    routine = new TestRoutine('key', 'title', { test: 123 });

    expect(routine.options).toEqual({ test: 123 });
  });

  it('creates an returns an `AggregatedPipeline`', () => {
    const context = new Context();
    const pipeline = routine.createAggregatedPipeline(context, 123);

    expect(pipeline).toBeInstanceOf(AggregatedPipeline);
    expect(pipeline.context).toBe(context);
    expect(pipeline.value).toBe(123);
  });

  it('creates an returns an `ConcurrentPipeline`', () => {
    const context = new Context();
    const pipeline = routine.createConcurrentPipeline(context, 123);

    expect(pipeline).toBeInstanceOf(ConcurrentPipeline);
    expect(pipeline.context).toBe(context);
    expect(pipeline.value).toBe(123);
  });

  it('creates an returns an `PooledPipeline`', () => {
    const context = new Context();
    const pipeline = routine.createPooledPipeline(context, 123);

    expect(pipeline).toBeInstanceOf(PooledPipeline);
    expect(pipeline.context).toBe(context);
    expect(pipeline.value).toBe(123);
  });

  it('creates an returns an `WaterfallPipeline`', () => {
    const context = new Context();
    const pipeline = routine.createWaterfallPipeline(context, 123);

    expect(pipeline).toBeInstanceOf(WaterfallPipeline);
    expect(pipeline.context).toBe(context);
    expect(pipeline.value).toBe(123);
  });

  describe('executeCommand()', () => {
    class FakeStream {
      pipe() {
        return this;
      }

      on(event: string, handler: (line: string) => any) {
        handler('Mocked stream line');

        return this;
      }

      toString() {
        return '';
      }
    }

    beforeEach(() => {
      ((execa as unknown) as jest.Mock).mockImplementation((command, args) => ({
        command: `${command} ${args.join(' ')}`,
        stdout: new FakeStream(),
        stderr: new FakeStream(),
      }));
    });

    it('runs a local command', async () => {
      const result = await routine.executeCommand('yarn', ['-v']);

      expect(result).toEqual(expect.objectContaining({ command: 'yarn -v' }));
    });

    it('runs a local command in a shell', async () => {
      const result = await routine.executeCommand('echo', ['boost'], { shell: true });

      expect(execa).toHaveBeenCalledWith('echo', ['boost'], { shell: true });
      expect(result).toEqual(expect.objectContaining({ command: 'echo boost' }));
    });

    it('pipes stdout/stderr to handler', async () => {
      const commandSpy = jest.fn();
      const commandDataSpy = jest.fn();
      const task = new Task('title', () => {});

      // @ts-ignore Allow
      task.status = STATUS_RUNNING;

      routine.onCommand.listen(commandSpy);
      routine.onCommandData.listen(commandDataSpy);

      await routine.executeCommand('yarn', ['-v'], { workUnit: task });

      expect(commandSpy).toHaveBeenCalledWith('yarn');
      expect(commandDataSpy).toHaveBeenCalledWith('yarn', expect.anything());
    });

    it('sets `statusText` and `output` on work unit', async () => {
      const task = new Task('title', () => {});

      // @ts-ignore Allow
      task.status = STATUS_RUNNING;

      await routine.executeCommand('yarn', ['--help'], { workUnit: task });

      expect(task.statusText).toBe('Mocked stream line');
      expect(task.output).toBe('Mocked stream lineMocked stream line');
    });

    it('doesnt set `statusText` or `output` on work unit when not running', async () => {
      const task = new Task('title', () => {});

      await routine.executeCommand('yarn', ['-v'], { workUnit: task });

      expect(task.statusText).toBe('');
      expect(task.output).toBe('');
    });
  });
});
