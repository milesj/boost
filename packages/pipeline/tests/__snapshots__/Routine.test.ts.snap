// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routine errors if key is not a string 1`] = `"Routine key must be a valid unique string. [PLN:ROUTINE_INVALID_KEY]"`;

exports[`Routine errors if no key is provided 1`] = `"Routine key must be a valid unique string. [PLN:ROUTINE_INVALID_KEY]"`;
